### PR TITLE
Mutiple Bugfixes

### PR DESCRIPTION
**Pull request by wxing1292**
* fix workflow timeout version check bug
* fix task event ID check bug
* fix integtest race condition

solve #779 #770

### CODE REVIEW

File: service/history/conflictResolver.go [hunk 1]
```diff
@@ -60,33 +62,44 @@ func (r *conflictResolver) reset(replayEventID int64) (*mutableStateBuilder, err
 	var resetMutableStateBuilder *mutableStateBuilder
 	var sBuilder *stateBuilder
 	requestID := uuid.New()
-	for hasMore := true; hasMore; hasMore = len(nextPageToken) > 0 {
-		history, nextPageToken, err = r.getHistory(domainID, execution, common.FirstEventID, replayNextEventID,
+
+	var lastFirstEventID int64
+	for remainingHistorySize := replayNextEventID - common.FirstEventID; remainingHistorySize > 0; {
```
[the remark this answers is not in the document]
Comment (wxing1292) [reply service/history/conflictResolver.go]: 1. i did not change the query
2. the change here make sure we are actually applying events up to the next event id. if the next event id < the highest event id in a batch of history

say in a batch, the first event id is 10, next event id is 15, and we only want events up to 13 (exclusive).
then this change can make it work, while before, we are actually applying all events from 10 -> 15

[the remark this answers is not in the document]
Comment (samarabbas) [reply service/history/conflictResolver.go]: My question is why can't we rely on nextPageToken for the looping condition.  I agree with you that last batch could have more events within the batch after nextEventID (although it is not possible unless we have a bug somewhere else).  But then it guarantees that there won't be any more batches after that.  So it is still ok to loop on the nextPageToken logic.

File: service/history/failoverCheck.go [hunk 2]
```diff
@@ -61,3 +64,77 @@ func verifyTimerTaskVersion(shard ShardContext, domainID string, version int64,
 	}
 	return true, nil
 }
+
+// load mutable state, if mutable state's next event ID <= task ID, will attempt to refresh
+// if still mutable state's next event ID <= task ID, will return nil, nil
+func loadMutableStateForTransferTask(context *workflowExecutionContext, transferTask *persistence.TransferTaskInfo, metricsClient metrics.Client, logger bark.Logger) (*mutableStateBuilder, error) {
```
Comment (samarabbas) [inline service/history/failoverCheck.go]: Why only transfer task?  Won't TimerTask have a similar issue?

File: service/history/failoverCheck.go [hunk 3]
```diff
@@ -61,3 +64,77 @@ func verifyTimerTaskVersion(shard ShardContext, domainID string, version int64,
...
+
+	// check to see if cache needs to be refreshed as we could potentially have stale workflow execution
+	// the exception is decision consistently fail
+	// there will be no event generated, thus making the decision schedule ID == next event ID
+	isDecisionRetry := transferTask.TaskType == persistence.TransferTaskTypeDecisionTask &&
+		msBuilder.executionInfo.DecisionScheduleID == transferTask.ScheduleID &&
+		msBuilder.executionInfo.DecisionAttempt > 0
+
+	if transferTask.ScheduleID >= msBuilder.GetNextEventID() && !isDecisionRetry {
+		metricsClient.IncCounter(metrics.TimerQueueProcessorScope, metrics.StaleMutableStateCounter)
```
Comment (samarabbas) [inline service/history/failoverCheck.go]: Probably you want to use a different metric scope here.

File: service/history/failoverCheck.go (hunk 2 above)
Comment (samarabbas) [reply service/history/failoverCheck.go]: Is there a way to generalize the check for both timer and transfer tasks?  Can we instead rely on version on the task and version on the mutable state?

Comment (wxing1292) [reply service/history/failoverCheck.go]: the main difference is here
```
isDecisionRetry := transferTask.TaskType == persistence.TransferTaskTypeDecisionTask &&
```
vs 
```
isDecisionRetry := timerTask.TaskType == persistence.TaskTypeDecisionTimeout &&
```

File: service/history/failoverCheck.go (hunk 3 above)
Comment (wxing1292) [reply service/history/failoverCheck.go]: fixed

**Pull request merged**